### DISCUSSION TITLE
storybook error

### DISCUSSION THREAD

**Discussion by wherehows** [amannn/next-intl#771] [Q&A]
### Description
I setup [Option 4](https://next-intl-docs.vercel.app/docs/environments/server-client-components#option-4-providing-all-messages) in my storybook and nextjs application

in nextjs application, the Option 4 setup works well. but in storybook, storybook makes an error even if i wrapping Story Component with `<NextIntlClientProvider>` as follow :

<img width="505" alt="CleanShot 2024-01-11 at 12 33 02@2x" src="https://github.com/amannn/next-intl/assets/81841082/58bbc260-c124-41b3-b794-f2cdd2ace2f1">

You can verify specific symptoms through CodeSandbox.

### Mandatory reproduction URL (CodeSandbox or GitHub repository)

https://codesandbox.io/p/devbox/next-intl-bug-template-app-forked-pxyx5n?file=%2F.storybook%2Fpreview.tsx%3A10%2C7&workspaceId=3b663f32-b4b0-43b5-9910-dc5ef7e82cb5

### Reproduction description

Steps to reproduce:
1. enter `yarn sb`in terminal
2. open 6006 port

### Expected behaviour

Storybook must not make an error

**Comment by amannn** [comment]
You're seeing this error because you're invoking `useMessages` in `preview.tsx` outside of `NextIntlClientProvider`. Here's [a fixed version](https://codesandbox.io/p/devbox/next-intl-bug-template-app-forked-kwkf3m?file=%2F.storybook%2Fpreview.tsx%3A24%2C6&workspaceId=90dbfd2c-4551-46fb-a76c-8f4b3bd87a46).

I'll move this to a discussion since it's a usage question.

**Reply by phibo23** [reply]
Hi, the code sandbox link doesn't work anymore. Is it the same solution as in the jest example where the static json is imported instead of useMessages? If it is a different solution, could you please repost the fixed example?
https://next-intl-docs.vercel.app/docs/faq#can-next-intl-be-used-with-jest-and-storybook
https://github.com/amannn/next-intl/blob/main/examples/example-app-router/src/components/Navigation.spec.tsx

**Reply by amannn** [reply]
You should be able to fork the Sandbox to your own workspace to see the code.

![Screenshot 2024-02-12 at 10 28 06](https://github.com/amannn/next-intl/assets/4038316/3fb13705-69ea-4310-997e-0f2fd0e0e403)

**Reply by phibo23** [reply]
I did try that, but the content seems to be unrelated to the storybook preview.tsx scenario that @wherehows described in his code sandbox https://codesandbox.io/p/devbox/next-intl-bug-template-app-forked-7w4p8n?workspaceId=2cc84ccc-52ef-43aa-8dee-0266375bb19e

**Reply by amannn** [reply]
This was the issue in that example:

> You're seeing this error because you're invoking `useMessages` in `preview.tsx` outside of `NextIntlClientProvider`.

The fix is to not use `useMessages` there but to import the messages via an import. `useMessages` really only returns what was configured on `NextIntlClientProvider` previously in that environment.

**Comment by Rikard-Johansson97** [comment]
I'm facing issues using getTranslations in Storybook 8.0 alpha with its new React Server Components (RSC) support.

[RSC PR with nextjs](https://github.com/storybookjs/storybook/pull/25263/files)

![image](https://github.com/amannn/next-intl/assets/106700975/24a16ad4-2d88-44d0-9f92-bb09b8f55bb5)

maybe its possible to mock getTranslation in some way.

```tsx

'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

/**
 * Allows to import `next-intl/server` in non-RSC environments.
 *
 * This is mostly relevant for testing, since e.g. a `generateMetadata`
 * export from a page might use `next-intl/server`, but the test
 * only uses the default export for a page.
 */

function notSupported(message) {
  return () => {
    throw new Error("`".concat(message, "` is not supported in Client Components."));
  };
}

```

**Reply by dmitrc** [reply]
I've encountered the same issue and here is how I fixed it for now:

* Create `.storybook/mocks/next-intl-server.ts`:
```
import { useTranslations } from "next-intl";

export const getTranslations = async (ns: string) => {
  return useTranslations(ns);
};
```

* Add in `webpackFinal` in `.storybook/main.ts`:
```
config.resolve ||= {};
config.resolve.alias ||= {};
config.resolve.alias = {
  ...config.resolve.alias,
  "next-intl/server": path.resolve(
    __dirname,
    "./mocks/next-intl-server.ts",
  ),
};
```

* Add a decorator in `.storybook/preview.tsx`:
```
import { NextIntlClientProvider } from "next-intl";
import en from "../src/messages/en.json";

const WithIntlProvider = (Story) => {
  return (
    <NextIntlClientProvider
      locale="en"
      messages={en}
    >
      <Story />
    </NextIntlClientProvider>
  );
};

const preview: Preview = {
  ...
  decorators: [WithIntlProvider],
}
```

In essence, do what Storybook itself does, and mock the server logic via the client functions.

It would be cool to expand it further to allow to select the locale w/out importing everything, but this will do it for now.
Cc @amannn if you have any better ideas on how to tackle this.

**Reply by amannn** [reply]
I haven't really experimented with this yet and am not familiar with how the RSC support in Storybook works. Generally, my advice would be to use [non-async components](https://next-intl-docs.vercel.app/docs/environments/server-client-components#shared-components) (aka "shared components") as much as possible, as this gives you the most flexibility (run in RSC, run in Storybook, run in unit tests, …).

That being said, from the bundler side `next-intl` uses the following features when ran as part of a Next.js app:
1. The `react-server` import condition is set
2. An alias for `next-intl/config` is set to `./src/i18n.ts` (done by the `next-intl` plugin in a regular setup)

If these two conditions are met, possibly Storybook could support running async components with `next-intl`.

To set a locale, maybe the newly added capability to return a `locale` from (a mock) `i18n.ts` could work (see [App Router setup without i18n routing: `i18n.ts`](https://next-intl-docs.vercel.app/docs/getting-started/app-router/without-i18n-routing#i18nts))

If anyone tests this, I'd be curious to hear if it works!

**Reply by loki344** [reply]
The solution provided by @dmitrc works for me, storybook version 8.1.4 and next-intl version 3.14.1. Only thing that does throw an exception are dynamic translations.
`Error: INVALID_MESSAGE: Cannot read properties of undefined (reading 'date')
    at getFallbackFromErrorAndNotify (VM2136216 vendors-node_modules_ant-design_nextjs-registry_es_index_js-node_modules_ant-design_pro-form_-dfa529.iframe.bundle.js:195866:19)
    at translateBaseFn (VM2136216 vendors-node_modules_ant-design_nextjs-registry_es_index_js-node_modules_ant-design_pro-form_-dfa529.iframe.bundle.js:195942:16)
    at translateFn (VM2136216 vendors-node_modules_ant-design_nextjs-registry_es_index_js-node_modules_ant-design_pro-form_-dfa529.iframe.bundle.js:195972:20)
    at VM2136286 src_components_business-object_table_BusinessObjectTable_tsx.iframe.bundle.js:8806:17
    at mountMemo (VM2136216 vendors-node_modules_ant-design_nextjs-registry_es_index_js-node_modules_ant-design_pro-form_-dfa529.iframe.bundle.js:127132:19)
    at Object.useMemo (VM2136216 vendors-node_modules_ant-design_nextjs-registry_es_index_js-node_modules_ant-design_pro-form_-dfa529.iframe.bundle.js:127837:16)
    at useMemo (VM2136216 vendors-node_modules_ant-design_nextjs-registry_es_index_js-node_modules_ant-design_pro-form_-dfa529.iframe.bundle.js:158578:21)
    at BusinessObjectTable (VM2136286 src_components_business-object_table_BusinessObjectTable_tsx.iframe.bundle.js:8803:93)
    at renderWithHooks (VM2136216 vendors-node_modules_ant-design_nextjs-registry_es_index_js-node_modules_ant-design_pro-form_-dfa529.iframe.bundle.js:125653:18)
    at mountIndeterminateComponent (VM2136216 vendors-node_modules_ant-design_nextjs-registry_es_index_js-node_modules_ant-design_pro-form_-dfa529.iframe.bundle.js:131401:13)`


I tracked it down to the use-intl library where the function convertFormatsToIntlMessageFormat accesses a undefined value.
![image](https://github.com/amannn/next-intl/assets/43857623/2ea7746c-5040-4ad2-9a4e-5843c79c5f51)

What seems strange is that the formats.date value would be available under IntlMessageFormat__default.default.default.formats.date (one default more).
![image](https://github.com/amannn/next-intl/assets/43857623/e90fe75b-9983-4e80-b1a5-f746c2dead5f)

preview.tsx

`                
//this import is important
import de from "../src/core/i18n/static-translations/messages/de.json";

// rest

return <NextAuthProvider session={session}>
                    <SessionProvider session={session}>
                        <NextIntlClientProvider
                            locale={'de'}
                            messages={de}
                            timeZone={'Europe/Zurich'}
                        >
                              <Suspense fallback={SSR_LOADING_TEXT}>
                                  <Story/>
                              </Suspense>
                        </NextIntlClientProvider>
                    </SessionProvider>
                </NextAuthProvider>`

**Reply by markomitranic** [reply]
Thank you all for the suggestions. I've had to proxy `getTranslations` and `getLocale` this way today, wouldn't have thought of this without you!

**Reply by loki344** [reply]
@markomitranic Would you mind showing how you proxied the methods?

**Reply by markomitranic** [reply]
@loki344 Sure, I did exactly as described above :) 

```ts
// .storybook/__mock__/next-intl-server.ts
import { useLocale, useTranslations } from "next-intl";

/**
 * `get` functions don't work in Storybook, so we mock them.
 * @see https://github.com/amannn/next-intl/discussions/771#discussioncomment-9587798
 */

export const getLocale = async () => useLocale();
export const getTranslations = async (
  ...args: Parameters<typeof useTranslations>
) => useTranslations(…args);
```

In order to do that, I had to add this file to Storybook's `main.ts` file, under webpack configuration:
```ts
const config: StorybookConfig = {
  //..... other stufff....

  webpackFinal: async (config, { configType }) => {
    config.resolve ||= {};
    config.resolve.alias ||= {};

    config.resolve.modules = [path.resolve(__dirname, ".."), "node_modules"];
    config.resolve.alias = {
      ...config.resolve.alias,

      // Allows for importing stuff like mocks from the .storybook folder.
      // @see https://github.com/storybookjs/storybook/issues/11639#issuecomment-690834754
      ".storybook": path.resolve(__dirname, "./"),

      // `getTranslations` doesn't work in Storybook, so we mock it.
      // @see https://github.com/amannn/next-intl/discussions/771#discussioncomment-9587798
      "next-intl/server": path.resolve(
        __dirname,
        "./__mock__/next-intl-server.ts",
      ),
    };

    return config;
  },
};

export default config;
```

Finally, I've had to ensure that my Storybook's `preview.tsx` understands `next-intl` as described in `next-intl` docs:

```ts
import { Preview } from "@storybook/react";
import { NextIntlClientProvider } from "next-intl";
import da from "../messages/da/messages.json";

const preview: Preview = {
   //..... other stufff....

  decorators: [
    (Story) => (
      <NextIntlClientProvider
        locale="da-DK"
        messages={da}
        timeZone={"Europe/Copenhagen"}
      >
        <div className="font-tex-regular">
          <Story />
        </div>
      </NextIntlClientProvider>
    ),
  ],
};

export default preview;
```

**UPDATE 25.06.2024:** Updated the code examples based on below comments.

**Reply by loki344** [reply]
@markomitranic thank you!
Are dynamic translations working for you?
e.g. if you have 
`"delete-title-single": "{item} löschen",`
is it translated correctly or does it simply show the key of the translation and throw an error in the log?

**Reply by dmitrc** [reply]
The above example is a very basic proof-of-concept. If you need additional arguments handled, you can replace `ns: string` part with `...args` and pass that onwards to `useTranslations` method.

**Reply by markomitranic** [reply]
Yes, I'm blind for not seeing it lol

```ts
export const getTranslations = async (
  ...args: Parameters<typeof useTranslations>
) => useTranslations(…args);
```

I'll update the above example for any future readers ;)

**Reply by Gawdfrey** [reply]
Thanks, you saved me a lot of time :)

**Reply by Demianeen** [reply]
@dmitrc, do you get Rules of Hook error when use use `useTranslations` like this?
![CleanShot 2025-02-19 at 11 39 52@2x](https://github.com/user-attachments/assets/0e3d2dce-e83f-4189-ade0-7fff58945056)

**Reply by dmitrc** [reply]
@Demianeen, depending on how strictly is your ESLint configured, you might need to suppress the warning:
```
export function getTranslations(...args: Parameters<typeof useTranslations>) {
  // eslint-disable-next-line react-hooks/rules-of-hooks
  return useTranslations(...args);
}
```

If there are still runtime errors afterwards, you should double-check your mocking setup and the translations usage (which should still be inside an RSC / function component).

**Comment by Gawdfrey** [comment]
Hmm this worked flawlessly, until I had pages that had two different `await getTranslations()` functions..
Getting `Error: Failed to call `useTranslations` because the context from `NextIntlClientProvider` was not found.`.
Removing the second one makes it render, and can't really understand why 😓 

Currently mocking it like this
```
import { useLocale, useMessages, useTranslations } from "next-intl";

/**
 * `get` functions don't work in Storybook, so we mock them.
 * @see https://github.com/amannn/next-intl/discussions/771#discussioncomment-9587798
 */

export const getLocale = async () => useLocale();
export const getTranslations = async (
  ...args: Parameters<typeof useTranslations>
) => useTranslations(args);
export const getMessages = async () => useMessages();

```

**Reply by alessiosferro** [reply]
Had the same issue, solved it by using Promise.all 😄

**Reply by Gawdfrey** [reply]
That worked perfectly, but still a bit annoying having to change real code for Storybook, but eh. Thanks for sharing!

**Reply by dmitrc** [reply]
This is actually a really interesting side effect, and I also was not able to get to the bottom of it yet.

With the simple mock above and proper `<NextIntlClientProvider>` wrapper, the story for this component works:
```js
export default async function Test() {
  const t = await getTranslations("Test");
  const obj = await somethingElse();
  return (...);
 }
 ```
 
 But this one doesn't:
 ```js
export default async function Test() {
  const obj = await somethingElse();
 
  // Failed to call `useTranslations` because the context from `NextIntlClientProvider` was not found.
  const t = await getTranslations("Test");

  return (...);
 }
```

Not really sure what makes the context get lost one layer deep here...
Probably the specifics of how Storybook emulates RSC components + potentially more weirdness.

Do you have any hunches for what could this possibly be, @amannn?

**Reply by dmitrc** [reply]
Started an issue over in the Storybook repo to look into this behavior further -- https://github.com/storybookjs/storybook/issues/30317. As of Storybook v8.5.0, the error about the missing context has now become an infinite re-render loop instead.
